### PR TITLE
Monitoring slow queries in MySQL with Stackdriver

### PR DESCRIPTION
**Pull request by jwlee98**
I've made Monitoring slow queries in MySQL with Stackdriver and requested review

@MrTrustor

### CODE REVIEW

**Comment by jwlee98** [comment]
@MrTrustor could you help me how to add reviewer ?

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 1]
```diff
@@ -0,0 +1,468 @@
+
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: This file should be name `index.md`.
Take a look at a recent tutorial to see the expect structure: https://github.com/GoogleCloudPlatform/community/tree/master/tutorials/understanding-oauth2-and-deploy-a-basic-auth-srv-to-cloud-functions

Also, this file should have a specific header: see an example here: https://raw.githubusercontent.com/GoogleCloudPlatform/community/master/tutorials/understanding-oauth2-and-deploy-a-basic-auth-srv-to-cloud-functions/index.md

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+The slow query log feature in [MySQL](http://mysql.com) enables you to log queries that exceed a predefined time limit. This greatly simplifies the task of finding inefficient or time-consuming queries. By integrating these logs with Stackdriver, you can keep an eye on database performance, allowing you to identify which queries need optimization and, optionally, be alerted when there are too many slow queries.
+
+
+![architecture diagram](stackdriver-monitor-slow-query-mysql/Copy-of0.jpg "image_tooltip")
+
+
+This solution can also be used with self-managed MySQL instances running on Compute Engine. See 
+
+See [Appendix A](#appendixa) for instructions on how to configure a self-managed instance to send slow query logs to Stackdriver.
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: You have 2 "Sees". One line 20 and one line 22.

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+### Objectives
+
+
+
+*   Learn how to enable slow query logs in Cloud SQL.
+*   Learn how to access slow query logs in Stackdriver Logging.
+*   Learn how to create log-based metrics for alerting and troubleshooting.
+
+
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+
+
+*   Cloud SQL for MySQL
+*   Stackdriver Logging
+*   Stackdriver Monitoring
+
+You can use the [pricing calculator](https://cloud.google.com/products/calculator/#id=817ac158-90f0-4e79-90f6-3e52c7b28a9c) to generate a cost estimate based on your projected usage. For example, if your slow query log contains 10 distinct SQL statements every minute, so they generate 0.033 MB of metrics per month, your total monthly cost is $0.00 if within the 150 MB free quota, or $0.78 if you have already exceeded 150 MB. For details, see this [pricing estimate](https://cloud.google.com/products/calculator/#id=6e6f8223-cbac-4ca4-8213-1778f8b35fb0).
+
+
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+
+1.  Browse to [Logs Viewer](https://console.cloud.google.com/logs/viewer).
+1.  From the leftmost dropdown, select **Cloud SQL** and the cloudsql001 database.
+1.  From the **All logs** dropdown, select **cloudsql.googleapis.com/mysql-slow.log**.
+
+
+![logviewer01](stackdriver-monitor-slow-query-mysql/Copy-of1.png "image_tooltip")
+
+
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+
+### <a name="generatesample">Generate sample data with a load test</as>
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not sure why you have a `<a ...></as>` here.

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+
+The test may take several minutes. Once it has completed, return to the Stackdriver Logging Cloud SQL view at https://console.cloud.google.com/logs/viewer. 
+
+
+![logviewer02](stackdriver-monitor-slow-query-mysql/Copy-of2.png "image_tooltip")
+
+
+Slow queries from the load test starting with "`SELECT"` should now be visible.
+
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+
+
+![metricexplorer](stackdriver-monitor-slow-query-mysql/Copy-of5.png "image_tooltip")
+
+
+
+
+You now have a metric that measures the number of slow query logs. You can use this to, for example, [trigger an alert](https://cloud.google.com/monitoring/alerts) if the number of slow queries exceeds a given threshold.
+
+The y-axis of the chart measures the number_ _of slow queries, not how long the slow queries took. To check the execution time of a single slow query, see the `Query_time` parameter in the slow log via Stackdriver Logging.  You can also create a [distribution metric](https://cloud.google.com/logging/docs/logs-based-metrics/distribution-metrics) on the regular expression `Query_time: ([0-9]+\.[0-9])`, which can be [visualized as a heatmap](https://cloud.google.com/logging/docs/logs-based-metrics/charts-and-alerts#create-chart).
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: There's a problem with underscores here. I think you can remove them.

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+
+
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+### Visualize slow query logs with Stackdriver monitoring
+
+The previously created metric measures the number of slow queries, but it does not show their content. You can make this metric more useful by parsing the slow SQL statements to be [metric labels](https://cloud.google.com/logging/docs/logs-based-metrics/labels), which are visible in the dashboard legend in Stackdriver.
+
+Return to the **User-Defined Metrics** view. Click the three dots icon to the left to open the menu, then select **Edit metric**. Choose **Add item** under **Labels** and enter the following values:
+
+Name: sql<br/>
+Label type: String<br/>
+Field name: textPayload<br/>
+Extraction regular expression: \b(SELECT.* WHERE|INSERT.* VALUES|UPDATE.* WHERE|CREATE.*|DELETE.*).*\z<br/>
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Bullet points for those 4 lines, and remove the `<br/>`

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+
+The dashboard now shows you slow MySQL queries, broken down by SQL statement. You can also hover over the graph to see queries per second rates for each statement. This allows you to quickly identify the culprit when alerted about database slowness. 
+
+
+
+![sqllogview](stackdriver-monitor-slow-query-mysql/Copy-of9.png "image_tooltip")
+
+
+&nbsp;
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+
+```
+gcloud sql instances delete cloudsql001
+gcloud logging metrics delete slow_sql
+rm -r ~/test_db
+```
+
+
+&nbsp;
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 14]
```diff
@@ -0,0 +1,468 @@
...
+
+
+
+## <a name="appendixa">[Appendix A]</a>: Configuring self-managed MySQL on Compute Engine to send slow query logs to Stackdriver
+
+The following instructions let you send slow query logs from a previously deployed self-managed MySQL instance running on Compute Engine to Stackdriver Logging.
+
+Note: The commands below assume that you already have a VM containing a self-managed MySQL database, and that the VM is running Debian/Ubuntu. Commands and file locations may vary in other Linux distributions. You can quickly deploy a new VM for test purposes by launching a [MySQL Google Click to Deploy instance](https://console.cloud.google.com/marketplace/details/click-to-deploy-images/mysql) from Marketplace.
+
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 15]
```diff
@@ -0,0 +1,468 @@
...
+
+
+Restart the google-fluentd agent to read the updated config:
+
+
+```
+sudo service google-fluentd restart
+```
+
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 16]
```diff
@@ -0,0 +1,468 @@
...
+```
+
+
+You have now configured MySQL and Stackdriver Logging for slow query logs. Proceed to 
+
+
+[Generate sample data with a load test](#generatesample).
+
+&nbsp;
+&nbsp;
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not needed

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 17]
```diff
@@ -0,0 +1,468 @@
...
+*   Learn how to access slow query logs in Stackdriver Logging.
+*   Learn how to create log-based metrics for alerting and troubleshooting.
+
+
+&nbsp;
+
+
+### Costs
+
+This tutorial uses the following billable components of Google Cloud Platform: \
```
Comment (MrTrustor) [inline tutorials/stackdriver-monitor-slow-query-mysql.md]: Not sure why you have a `\` here.

File: tutorials/stackdriver-monitor-slow-query-mysql.md [hunk 18]
```diff
@@ -0,0 +1,468 @@
+
+## Monitoring slow queries in MySQL with Stackdriver
+
+<p style="text-align: right">
+jungwoonlee@, jani@ \
+<a href="https://jira.gcp.solutions/browse/PUB-1130">https://jira.gcp.solutions/browse/PUB-1130</a></p>
```
[the remark this answers is not in the document]
Comment (jwlee98) [reply tutorials/stackdriver-monitor-slow-query-mysql.md]: I removed it according to your guide

File: tutorials/stackdriver-monitor-slow-query-mysql.md (hunk 1 above)
Comment (jwlee98) [reply tutorials/stackdriver-monitor-slow-query-mysql.md]: I've change it

File: tutorials/stackdriver-monitor-slow-query-mysql.md (hunk 3 above)
Comment (jwlee98) [reply tutorials/stackdriver-monitor-slow-query-mysql.md]: I removed it according to your guide

File: tutorials/stackdriver-monitor-slow-query-mysql.md (hunk 4 above)
Comment (jwlee98) [reply tutorials/stackdriver-monitor-slow-query-mysql.md]: I removed it according to your guide

File: tutorials/stackdriver-monitor-slow-query-mysql.md (hunk 13 above)
Comment (jwlee98) [reply tutorials/stackdriver-monitor-slow-query-mysql.md]: I removed it according to your guide

File: tutorials/stackdriver-monitor-slow-query-mysql.md (hunk 14 above)
Comment (jwlee98) [reply tutorials/stackdriver-monitor-slow-query-mysql.md]: I removed it according to your guide

File: tutorials/stackdriver-monitor-slow-query-mysql.md (hunk 16 above)
Comment (jwlee98) [reply tutorials/stackdriver-monitor-slow-query-mysql.md]: I removed it according to your guide

File: tutorials/stackdriver-monitor-slow-query-mysql.md (hunk 17 above)
Comment (jwlee98) [reply tutorials/stackdriver-monitor-slow-query-mysql.md]: I removed it according to your guide

File: tutorials/stackdriver-monitor-slow-query-mysql.md (hunk 15 above)
Comment (jwlee98) [reply tutorials/stackdriver-monitor-slow-query-mysql.md]: I removed it according to your guide

**Comment by MrTrustor** [comment]
Looking better. We probably need to wait for next week for further progress on this (for @ccarpentiere or someone on her team to take a look).
Especially, I'd like to know if we can preview the result somewhere.

**Comment by jwlee98** [comment]
Got it. Thanks for your kind support.

**Comment by ToddKopriva** [comment]
I’ll help with editorial and product review on Monday.

**Comment by LiveChief** [comment]
test

**Pull request merged**